### PR TITLE
cmd/extract-notes: Implement note extraction per version

### PR DESCRIPTION
**Pull request by radeksimko**
## Use case

Some projects may not be at such a scale that they would want or need to adopt any changelog generation mechanism. They may however still benefit from some automation when uploading the relevant changelog at release time, for example to GitHub Releases.

The output is _assumed_ to match the output of `changelog-build`, given the default template and default regular expression in the parser.

## Implementation Details

The parser was designed with some other future use cases in mind - e.g.

 - `SectionRange().BodyRange` can be used to identify the range where notes are and replace the relevant section of changelog; This allows regenerating an "unreleased" changelog straight after merging a PR.
 - `SectionRange().HeaderRange` can be used to identify the range of header and replace `(Unreleased)` with a release date
   - possibly even automate Changelog updates in [Vagrant's style](https://github.com/hashicorp/vagrant/blob/master/CHANGELOG.md#next-version-unreleased) - i.e. replace `Next version (Unreleased)` with `VERSION (RELEASE-DATE)`

The parser is as accurate as the regular expression that drives it, but I think the default one is flexible enough (basically just assumes valid Markdown `##` header, no particular versioning) to cover format that most HashiCorp projects use for their Changelogs.

## Usage

```
terraform-provider-google $ changelog-extract-notes -version 3.15.0
```
```
FEATURES:
...
IMPROVEMENTS:
...
BUG FIXES:
...
```

This also allows easy plugging into [GoReleaser's interface](https://goreleaser.com/customization/#custom-release-notes), e.g.

```
$ goreleaser --release-notes <(changelog-extract-notes -version $GIT_TAG)
```

### CODE REVIEW

File: cmd/changelog-extract-notes/main.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	f, err := os.Open(changelogPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to open changelog file: %s", err)
+		os.Exit(1)
+	}
+
+	b, err := ioutil.ReadAll(f)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to read changelog file: %s", err)
+		os.Exit(1)
```
Comment (paddycarver) [inline cmd/changelog-extract-notes/main.go:40]: Could we replace this with [`ioutil.ReadFile`](https://golang.org/pkg/io/ioutil/#ReadFile)?

File: cmd/changelog-extract-notes/main.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	sp := parser.NewSectionParser(b)
```
Comment (paddycarver) [inline cmd/changelog-extract-notes/main.go:43]: Does it make sense for this to be an `io.Reader`?

File: cmd/changelog-extract-notes/main.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+func main() {
+	wd, err := os.Getwd()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	var extractVersion, changelogPath string
+	flag.StringVar(&changelogPath, "path", filepath.Join(wd, "CHANGELOG.md"), "path to the changelog file")
+	flag.StringVar(&extractVersion, "version", "", "version to extract changelog for (e.g. 1.0.0)")
```
Comment (paddycarver) [inline cmd/changelog-extract-notes/main.go:22]: What if this was the first argument instead of a flag? We only have the one required option, so I don't know that the flag gives it any naming benefits, and I think making it more visible as an argument instead of a flag may be nice?

File: parser/errors.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package parser
+
+import "fmt"
+
+type VersionNotFoundErr struct {
+	Version string
+}
+
+func (e *VersionNotFoundErr) Is(target error) bool {
```
Comment (paddycarver) [inline parser/errors.go:9]: Is `IsVersion` a better name for this method? `Is` isn't immediately clear to me in terms of what it does. Or is this meant to be an equality check? In which case, is `Equal` better, a la `time.Equal`?

File: parser/parser.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+)
+
+var (
+	defaultSectionReFmt = `(?s)(?P<header>## %s[^\n]*)
+(?P<body>.+?)
+(## .+|$)`
```
Comment (paddycarver) [inline parser/parser.go:12]: This could be `(?:## .+|$)`, which would create a non-capturing group, to not pollute the matches.

**Comment by radeksimko** [comment]
I will respond to each inline comment individually, but to your comment about regex - I understand and agree.

My choice of regex came purely from the perspective of it being universal enough to deal with any changelog format. I assumed that universal changelog format was something we want to support by looking at the logic which _generates_ changelog - i.e. I don't see any Markdown specific logic - just plain text templates that can contain anything, even invalid Markdown (for better or worse).

In other words this was a practical choice of trying to provide parity/compatibility with the current state of the generating part of go-changelog.

If we do go down the road of the parser understanding/prescribing certain formats like Markdown - which I think is a good idea and makes the whole thing less fragile and more maintainable - I think we should first ensure that we're actually _generating_ a valid and parseable changelog. That doesn't mean we have to be generating changelog dynamically and abandon the `text/template`, but we should at least run the output through the same markdown parser which would be then parsing it to validate that it's actually parseable.

TL;DR I went for consistency, rather than best solution here.

**Comment by radeksimko** [comment]
> My thoughts on that are nascent, but it feels like a configurable "section break" token could allow us to split into sections (in this example, it may be `\n## `), a "header break" token could allow us to split into header and body (in this case, it could just be the first \n in the section) and then a "version" token could allow us to parse the version out (in this case, it could be the first v in the header).

I like the section break token idea, but I'm not sure whether parsing the header/version the way you're suggesting would be accurate enough. I would probably still advocate for a regexp pattern there, which should however be much more readable as it would only need to match against a single line.

So perhaps to maintain the balance between fragility, readability and accuracy we need to leverage both real Markdown parser and regexp? 🤔

File: cmd/changelog-extract-notes/main.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+		os.Exit(1)
+	}
+
+	f, err := os.Open(changelogPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "unable to open changelog file: %s", err)
+		os.Exit(1)
+	}
+
+	b, err := ioutil.ReadAll(f)
```
Comment (paddycarver) [inline cmd/changelog-extract-notes/main.go:37]:
```suggestion
	b, err := ioutil.ReadFile(changelogPath)
```

File: cmd/changelog-extract-notes/main.go (hunk 2 above)
Comment (paddycarver) [reply cmd/changelog-extract-notes/main.go:43]: Sorry, to be clear, I meant does it make sense for `parser.NewSectionParser` to accept an `io.Reader`, not for it to return an `io.Reader`.

File: parser/parser.go (hunk 5 above)
Comment (paddycarver) [reply parser/parser.go:12]:
```suggestion
(?:## .+|$)`
```

File: cmd/changelog-extract-notes/main.go (hunk 3 above)
Comment (paddycarver) [reply cmd/changelog-extract-notes/main.go:22]:
```suggestion
	extractVersion := flag.Arg(0)
```

File: cmd/changelog-extract-notes/main.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+func main() {
+	wd, err := os.Getwd()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+
+	var extractVersion, changelogPath string
```
Comment (paddycarver) [inline cmd/changelog-extract-notes/main.go:20]:
```suggestion
	var changelogPath string
```

**Comment by bflad** [comment]
This would be super helpful to remove shell scripting workarounds like:

```
sed -n -e "1{/# /d;}" -e "2{/^$/d;}" -e "/# $(git describe --abbrev=0 --exclude="$(git describe --abbrev=0 --match='v*.*.*' --tags)" --match='v*.*.*' --tags | tr -d v)/q;p" CHANGELOG.md > /tmp/RELEASE-NOTES.md
```

File: cmd/changelog-extract-notes/main.go (hunk 2 above)
Comment (radeksimko) [reply cmd/changelog-extract-notes/main.go:39]: I suppose it's better in terms of the "outside" contract but I'm not sure whether it's a good idea to entirely avoid the byte slice "cache" inside the parser given that we'd then end up reading _everything_ from the reader at least twice, or have to implement some sort of cache-on-first-read, which feels rather overcomplicated.

File: parser/errors.go (hunk 4 above)
Comment (radeksimko) [reply parser/errors.go:9]: It is not, because this is meant to implement/satisfy the `errors.Is` interface https://pkg.go.dev/errors#Is
